src/nvim/eval/typval.c

```diff
@@ -0,0 +1,2556 @@
+///                   Only used when `deep` is true.
+/// @param[in]  orig  Original list to copy.
+/// @param[in]  deep  If false, then shallow copy will be done.
+/// @param[in]  copyID  See var_item_copy().
+///
+/// @return Copied list. May be NULL in case original list is NULL or some
+///         failure happens. The refcount of the new list is set to 1.
+list_T *tv_list_copy(const vimconv_T *const conv, list_T *const orig,
+                     const bool deep, const int copyID)
+  FUNC_ATTR_WARN_UNUSED_RESULT
+{
+  if (orig == NULL) {
+    return NULL;
+  }
+
+  list_T *copy = tv_list_alloc();
+  if (copyID != 0) {
+    // Do this before adding the items, because one of the items may
+    // refer back to this list.
+    orig->lv_copyID = copyID;
+    orig->lv_copylist = copy;
+  }
+  listitem_T *item;
+  for (item = orig->lv_first; item != NULL && !got_int;
+       item = item->li_next) {
+    listitem_T *const ni = tv_list_item_alloc();
+    if (deep) {
+      if (var_item_copy(conv, &item->li_tv, &ni->li_tv, deep, copyID) == FAIL) {
+        xfree(ni);
+        break;
+      }
+    } else {
+      tv_copy(&item->li_tv, &ni->li_tv);
+    }
+    tv_list_append(copy, ni);
+  }
+  copy->lv_refcount++;
+  if (item != NULL) {
+    tv_list_unref(copy);
+    copy = NULL;
+  }
+
+  return copy;
+}
+
+/// Extend first list with the second
+///
+/// @param[out]  l1  List to extend.
+/// @param[in]  l2  List to extend with.
+/// @param[in]  bef  If not NULL, extends before this item.
+void tv_list_extend(list_T *const l1, list_T *const l2,
+                    listitem_T *const bef)
+  FUNC_ATTR_NONNULL_ARG(1, 2)
+{
+  int todo = l2->lv_len;
+  listitem_T *const befbef = (bef == NULL ? NULL : bef->li_prev);
+  listitem_T *const saved_next = (befbef == NULL ? NULL : befbef->li_next);
+  // We also quit the loop when we have inserted the original item count of
+  // the list, avoid a hang when we extend a list with itself.
+  for (listitem_T *item = l2->lv_first
+       ; item != NULL && --todo >= 0
+       ; item = (item == befbef ? saved_next : item->li_next)) {
+    tv_list_insert_tv(l1, &item->li_tv, bef);
+  }
+}
+
+/// Concatenate lists into a new list
+///
+/// @param[in]  l1  First list.
+/// @param[in]  l2  Second list.
+/// @param[out]  ret_tv  Location where new list is saved.
+///
+/// @return OK or FAIL.
+int tv_list_concat(list_T *const l1, list_T *const l2, typval_T *const tv)
+  FUNC_ATTR_WARN_UNUSED_RESULT
+{
+  list_T *l;
+
+  tv->v_type = VAR_LIST;
+
+  if (l1 == NULL && l2 == NULL) {
+    l = NULL;
+  } else if (l1 == NULL) {
+    l = tv_list_copy(NULL, l2, false, 0);
+  } else {
+    l = tv_list_copy(NULL, l1, false, 0);
+    if (l != NULL && l2 != NULL) {
+      tv_list_extend(l, l2, NULL);
+    }
+  }
+  if (l == NULL && !(l1 == NULL && l2 == NULL)) {
+    return FAIL;
+  }
+
+  tv->vval.v_list = l;
+  return OK;
+}
+
+typedef struct {
+  char_u *s;
+  char_u *tofree;
```

Comment:
Why is `tofree` needed? Can't we free with `s` too?

To optimize some `strlen()` calls out, we could save the length of `s` here too and later use `ga_concat_len()` instead of `ga_concat()` in `list_join_inner()`.